<h1>Tabla de contenido<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#PROGRAMACIÓN-ORIENTADA-A-OBJETOS" data-toc-modified-id="PROGRAMACIÓN-ORIENTADA-A-OBJETOS-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>PROGRAMACIÓN ORIENTADA A OBJETOS</a></span><ul class="toc-item"><li><span><a href="#Introducción" data-toc-modified-id="Introducción-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Introducción</a></span></li><li><span><a href="#Resumiendo" data-toc-modified-id="Resumiendo-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Resumiendo</a></span><ul class="toc-item"><li><span><a href="#Clase" data-toc-modified-id="Clase-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Clase</a></span></li><li><span><a href="#Metodo" data-toc-modified-id="Metodo-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Metodo</a></span></li><li><span><a href="#Objeto" data-toc-modified-id="Objeto-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>Objeto</a></span></li><li><span><a href="#Instanciar" data-toc-modified-id="Instanciar-1.2.4"><span class="toc-item-num">1.2.4&nbsp;&nbsp;</span>Instanciar</a></span></li></ul></li><li><span><a href="#Ejemplo-Práctico" data-toc-modified-id="Ejemplo-Práctico-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Ejemplo Práctico</a></span></li></ul></li><li><span><a href="#Ejercicio-en-Python" data-toc-modified-id="Ejercicio-en-Python-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Ejercicio en Python</a></span></li><li><span><a href="#Ejercicio-en-Word" data-toc-modified-id="Ejercicio-en-Word-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ejercicio en Word</a></span></li></ul></div>

# PROGRAMACIÓN ORIENTADA A OBJETOS

 

## Introducción
> Imagine una clase como un conjunto de elementos de la vida natural, que comparten características compartidas, por ejemplo, __Animales mamíferos__, esta clase involucra un muy amplio numero de especies y éstas a su vez de subespecies, los animales mamíferos tienen una serie de características comunes con un valor para cada especie, incluso distintos, pero siguen siendo mamíferos, por ejemplo: nombre, edad, color, localización, tipo de alimentación, hábitat, tiempo de gestación, tiempo de destete, etc. A estas características se le conocen como __Atributos__. Los atributos son datos con valores usados para describir los integrantes de una clase. Muchos de estos atributos no contienen valores constantes, ya que pueden cambiar con el tiempo, por ejemplo: la edad, vida fértil, localización. Por lo tanto, muchas veces es necesario realizar operaciones con los atributos o con otras variables de la clase para actualizarlos o modificarlos, éstas operaciones se almacenan en funciones especiales llamadas __métodos__.  
>
> Piense  en un animal pertenciente a la clase _mamíferos,_ como el  gato doméstico (_Felis silvestris catus),_ éste como todos los demás animales de la misma clase, no importa la especie, se les conoce en la __POO__ como __Objetos__.
>
> La acción de crear un objeto para una clase en específico se le conococe como __instanciar__, y el objeto como tal, es una __instancia de clase__.
>
> Las clases se representan en __POO__ mediante Diagramas de clases, en el estándar UML Lenguaje unificado de modelado _(UML, por sus siglas en inglés, Unified Modeling Language),_ siguiendo con el ejemplo de animales:
> ![DIAGRAMA_DE_CLASES.png](attachment:DIAGRAMA_DE_CLASES.png)
>
> Observe como aparte de la clase Mamíferos, se desprenden otras subclases (gatos, perros, elefantes), éstas como conservan los mismos atributos de la clase principal no requieren ser nombrados nuevamente en el diagrama, por lo tanto se dice que tienen una relación de __herencia__ con la clase _animales._ 
>
> Los signos "+" y "-" al inicio de cada atributo o método indican la visibiliad de estos, indicando si son públicos o privados respectivamente.

## Resumiendo
Practicamente todo el codigo de Python realiza acciones con objetos.
### Clase
> Es un conjunto de elementos con caracteristicas compartidas que se expresan con: propiedades, variables y funciones”. Las clases definen conjuntos de atributos que se pueden compartir en todas las instancias de la clase.
### Metodo
> Las funciones definidas dentro de una clase  se conocen como metodos, un __metodo de instancia__ son aquellas funciones que operan sobre una instancia de una clase (objeto), no todos los metodos son _metodos de instancia_. La forma de indicar esto es atraves de un argumento llamado __self__ el cual indica que un objeto que se instancie puede usar esos metodos de instancia, por que de cierta manera "son propios de ese objeto" o "le pertenecen". pero cualquier identificador legal puede usarse para ello.
### Objeto
> Es un elemento que pertenece a una clase, se dice que un objeto es una instancia de una clase. La palabra reservada __Class__ no crea por si misma una instancia de una clase.
### Instanciar
> Es la accion de crear un objeto que pertenezca a una clase en especifico.

## Ejemplo Práctico

>Un buen ejemplo para poder entender el funcionamiento de una clase es considerar el caso de una biblioteca, ya que podemos facilmente asociar los elementos que pertenecen a ese conjunto, en este caso los objetos serian los libros 
>
> Se inicia creando una la clase _Library:_

In [97]:
class empleado(object):
    numEmpleado = 0
    def __init__(self, nombre, bonificacion):
        self.abono = 0
        self.nombre = nombre

        self.bonificacion = bonificacion
        empleado.numEmpleado += 1
    def __del__(self):
        empleado.numEmpleado -= 1
    def horas(self, numHoras):
        self.abono += numHoras*self.bonificacion
        return ("{:04.2f} horas trabajadas").format(numHoras)
    def pago(self):
        self.abono = 0
        return ("pagado a {}").format(self.nombre)

>La primera línea crea la clase, incluye la variable numEmpleado que se irá auto incrementando conforme se agregan nuevos empleados. En este momento el empleado no tiene abonos, por esto su valor es cero. El resto debe ser la declaración de cada uno de los argumentos (nombre y bonificación).
>
>El método "del" elimina un emplado, el método "horas" calcula el número de horas totales que corresponde a las trabajadas multiplicadas por la bonificación, por ejemplo si se reconoce el 15%, la bondicacion seria de 1.15. El formato _:04.2f_ significa que se mostrarán un total de 4 digitos, de los cuales 2 serán decimales (float). 
>
> Este método muestra un mensaje indicando el proceso actual o final en que se encuentra un empleado.
>
>Una __variable de clase,__ como _numEmpleado,_ compartirá sus valores con todas las instancias de la clase. En el ejemplo, usamos numEmpleado para contar cuántas instancias de empleados hay. Nota: la clase llamada _Empleado_ implementará los métodos especiales llamados _init_ y _del_ . Primero, veamos un ejemplo para crear una instancia del objeto Empleado, cómo ejecutar los métodos y cómo devolver la clase y las variables de instancia.
>
>Agreguemos a dos empleados Juan y Luis , los cuales serán identificados con las variables emp1 y emp2 respectivamente, su bonificación será del 33% y 15% según el caso:

In [100]:
empA = empleado("Juan", 1.33)
empB = empleado("Luis", 1.15)

> A partir de este momento todos los métodos de la clase empleado pasarán a los objetos emp1 y emp2, en este punto si queremos saber cuantos empleados hay en total, llamammos la clase y el método correspindiente:

In [101]:
print(empleado.numEmpleado)

2


> Entrega como resultado 2, puesto que hemos agrefado a Juan y Luis.
>
> Suponga ahora desea registrarle a Juan 20 horas de trabajo, solo basta llamar el método "horas":

In [102]:
empA.horas(20)

'20.00 horas trabajadas'

> Para obtener el numero de horas trabajadas con la bonificación se procede igual, pero llamando el método "pago":

In [103]:
empA.abono

26.6

> Efectivamente, debido a que la bonoficación de Juan es de 1.33:
>
> $ Bonif.Juan = 20 x 1.33 = 26.6 $

> Para finalizar el proceso de Juan solo basta con llamar el metodo "Pago":

In [104]:
empA.pago()

'pagado a Juan'

> Para eliminar a Juan se llama el metodo "Del":

In [105]:
empA.__del__()
print(empleado.numEmpleado)

1


> El proceso está correcto puesto que efectivamente queda un solo empleado.

# Ejercicio en Python

>Debe realizarlo en Python, puede descargar la versión 3.7 desde acá:
>
> https://www.python.org/ftp/python/3.7.7/python-3.7.7-amd64.exe
>
>O si lo prefiere use un editor en línea, donde puede trabajar y descargar su archivo (no necesita instalar nada):
>
>https://repl.it/languages/python3

1. En Python  cree dos clases de su interés, cada una debe tener al menos 4 métodos  y 1 variable de clase, estos pueden hacer diferentes tareas como mostrar un mensaje y realizar alguna operación matemática. Instancie varios objetos a la clase y pruebe los diferentes métodos.
2. Guarde su archivo con el siguiente formato: __apellido_curso_guia3.py__

# Ejercicio en Word
Conteste las siguientes preguntas en un archivo de Word:

1. Que es una instancia de clase
2. Defina que son los metodos especiales de clase en programación orientada a objetos
3. Indique a qué hace referencia el término: Encapsulamiento de datos
4. Cite las fuentes bibliográficas donde obtuvo la información, haciendo uso de los conceptos aprendidos en el primer periodo.
5. Agregue la bibliografía siguiendo las indicaciones vistas en el primer periodo.
6. Guarde su archivo como: __apellido_curso_guia3teoria__